### PR TITLE
MSEARCH-793: Make DB Changes for Reindex Merge

### PR DESCRIPTION
**Pull request by mukhiddin-yusuf**
### Purpose
_Make DB Changes for Reindex Merge_

### Approach
_Create migration scripts to create DB tables, view and triggers_

### Changes Checklist
- [ ] **API Changes**: Document any API paths, methods, request or response bodies changed, added, or removed.
- [ ] **Database Schema Changes**: Indicate any database schema changes and their impact. Confirm that migration scripts were created.
- [ ] **Interface Version Changes**: Indicate any changes to interface versions.
- [ ] **Interface Dependencies**: Document added or removed dependencies.
- [ ] **Permissions**: Document any changes to permissions.
- [ ] **Logging**: Confirm that logging is appropriately handled.
- [ ] **Unit Testing**: Confirm that changed classes were covered by unit tests.
- [ ] **Integration Testing**: Confirm that changed logic was covered by integration tests.
- [x] **Manual Testing**: Confirm that changes were tested on local or dev environment.
- [ ] **NEWS**: Confirm that the NEWS file is updated with relevant information about the changes made in this pull request.

### Related Issues
[MSEARCH-793](https://folio-org.atlassian.net/browse/MSEARCH-793)

### CODE REVIEW

**Comment by viacheslavkol** [comment]
Don't we need to fill range tables right away?

**Comment by mukhiddin-yusuf** [comment]
> Don't we need to fill range tables right away?

if I get it correctly, then no, it would be handled separately in https://folio-org.atlassian.net/browse/MSEARCH-794 or https://folio-org.atlassian.net/browse/MSEARCH-799

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    <preConditions>
+      <not>
+        <tableExists tableName="instances"/>
+      </not>
+    </preConditions>
+
+    <comment>Create instances table</comment>
+
+    <createTable tableName="instances">
+      <column name="instance_id" type="VARCHAR(255)">
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:18]: Let's make it UUID type

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+  <changeSet id="MSEARCH-793@@create-instances-table" author="Mukhiddin_Yusupov">
+    <preConditions>
+      <not>
+        <tableExists tableName="instances"/>
+      </not>
+    </preConditions>
+
+    <comment>Create instances table</comment>
+
+    <createTable tableName="instances">
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:17]: It's better to name DB tables in singular

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+  <changeSet id="MSEARCH-793@@create-holdings-table" author="Mukhiddin_Yusupov">
+    <preConditions>
+      <not>
+        <tableExists tableName="holdings"/>
+      </not>
+    </preConditions>
+
+    <comment>Create holdings table</comment>
+
+    <createTable tableName="holdings">
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:45]: Singular and UUIDs as PK and FK

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+  <changeSet id="MSEARCH-793@@create-items-table" author="Mukhiddin_Yusupov">
+    <preConditions>
+      <not>
+        <tableExists tableName="items"/>
+      </not>
+    </preConditions>
+
+    <comment>Create items table</comment>
+
+    <createTable tableName="items">
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:84]: Singular and UUIDs as PK and FK

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      <column name="instance_id" type="VARCHAR(255)">
+        <constraints nullable="false"/>
+      </column>
+      <column name="holding_json" type="jsonb">
+        <constraints nullable="false"/>
+      </column>
+    </createTable>
+  </changeSet>
+
+  <changeSet id="MSEARCH-793@@create-idx_holdings_instance_id" author="Mukhiddin_Yusupov">
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:61]: Let's move creating indexes into creating table changes to make it one transaction.

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  <changeSet id="MSEARCH-793@@create-classifications-table" author="Mukhiddin_Yusupov">
+    <preConditions>
+      <not>
+        <tableExists tableName="classifications"/>
+      </not>
+    </preConditions>
+
+    <comment>Create classifications table</comment>
+
+    <createTable tableName="classifications">
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:123]: Add additional id PK column. We should generate it similar to subjects and contributors.

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    <preConditions>
+      <not>
+        <tableExists tableName="subjects"/>
+      </not>
+    </preConditions>
+
+    <comment>Create subjects table</comment>
+
+    <createTable tableName="subjects">
+      <column name="subject_id" type="VARCHAR(255)">
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:146]: Let's decrease amount of chars for subject_id and for auhtority_id

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+      <column name="contributor_name" type="VARCHAR(255)">
+        <constraints nullable="false" primaryKey="true" primaryKeyName="pk_contributors"/>
+      </column>
+      <column name="authority_id" type="VARCHAR(255)" defaultValue="">
+        <constraints nullable="false"/>
+      </column>
+      <column name="instances_arr" type="jsonb">
+        <constraints nullable="false"/>
+      </column>
+    </createTable>
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:183]: It should follow same structure as the index: 
```
"name": {
      "index": "keyword_lowercase",
      "showInResponse": [ "browse" ]
    },
    "contributorNameTypeId": {
      "index": "keyword",
      "searchTypes": [ "facet", "filter" ],
      "showInResponse": [ "browse" ]
    },
    "instances": {
      "type": "object",
      "properties": {
        "instanceId": {
          "index": "source"
        },
        "contributorTypeId": {
          "index": "keyword",
          "showInResponse": [ "browse" ]
        },
        "tenantId": {
          "index": "keyword",
          "searchTypes": [ "facet", "filter" ]
        },
        "shared": {
          "index": "bool",
          "searchTypes": [ "facet", "filter" ],
          "default": false
        }
      }
    },
    "authorityId": {
      "index": "keyword",
      "showInResponse": [ "browse" ]
    }
  }
```
`contributorNameTypeId` column is missing. And also please check trigger

File: src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    </preConditions>
+
+    <comment>Create merge_ranges table</comment>
+
+    <createTable tableName="merge_ranges">
+      <column name="entity_type" type="VARCHAR(255)">
+        <constraints nullable="false" primaryKey="true" primaryKeyName="pk_merge_ranges"/>
+      </column>
+      <column name="tenant_id" type="VARCHAR(255)">
+        <constraints nullable="false" primaryKey="true" primaryKeyName="pk_merge_ranges"/>
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v3.3/create-reindex-entity-tables.xml:200]: Let's decrease size

File: src/main/resources/changelog/changelog-master.xml [hunk 10]
```diff
@@ -9,4 +9,5 @@
   <include file="changes/v3.0/create_consortium_instance.xml" relativeToChangelogFile="true"/>
   <include file="changes/v3.2/create_instance_classification_table.xml" relativeToChangelogFile="true"/>
   <include file="changes/v3.2/create_browse_config_table.xml" relativeToChangelogFile="true"/>
+  <include file="changes/v3.3/create-reindex-entity-tables.xml" relativeToChangelogFile="true"/>
```
Comment (psmagin) [inline src/main/resources/changelog/changelog-master.xml:12]: Let's have v4.0 and update the version in pom.xml

**Review by psmagin — CHANGES_REQUESTED** [review]

File: src/main/resources/changelog/changes/v4.0/create-instance-trigger.sql [hunk 11]
```diff
@@ -6,61 +6,65 @@ BEGIN
...
+    ON CONFLICT (id)
+      DO UPDATE SET instances = subject.instances || excluded.instances;
   END LOOP;
 
   -- extract contributors
   FOR i IN SELECT * FROM jsonb_array_elements(NEW.instance_json -> 'contributors') LOOP
     INSERT
-    INTO contributors(contributor_id, contributor_name, instances_arr)
+    INTO contributor(id, name, contributorNameTypeId, authorityId, instances)
     VALUES (encode(digest((i ->> 'name' || i ->> 'authorityId')::bytea, 'sha1'), 'hex'), -- requires pgcrypto enabled
```
Comment (psmagin) [inline src/main/resources/changelog/changes/v4.0/create-instance-trigger.sql:46]: Id should include contributorNameTypeId

**Review by psmagin — APPROVED** [review]

**Review by viacheslavpoliakov — APPROVED** [review]

**Pull request merged**